src/deprecated-macros.hpp

```diff
@@ -0,0 +1,48 @@
+//
+// Copyright (c) 2018 INRIA
+//
+// This file is part of Pinocchio
+// Pinocchio is free software: you can redistribute it
+// and/or modify it under the terms of the GNU Lesser General Public
+// License as published by the Free Software Foundation, either version
+// 3 of the License, or (at your option) any later version.
+//
+// Pinocchio is distributed in the hope that it will be
+// useful, but WITHOUT ANY WARRANTY; without even the implied warranty
+// of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+// General Lesser Public License for more details. You should have
+// received a copy of the GNU Lesser General Public License along with
+// Pinocchio If not, see
+// <http://www.gnu.org/licenses/>.
+
+#ifndef __se3_deprecated_macros_hpp__
+#define __se3_deprecated_macros_hpp__
+
+#ifdef PINOCCHIO_WITH_HPP_FCL
+  #ifndef WITH_HPP_FCL
+    #define WITH_HPP_FCL // for backward compatibility
+  #endif
+  #ifndef PINOCCHIO_DISABLE_DEPRECATED_MACRO_WARNING
+    #pragma message("'WITH_HPP_FCL' is deprecated. Please use PINOCCHIO_WITH_HPP_FCL instead.")
+  #endif
```

Comment:
Is this way of handling the deprecation correct? It seems the warning will be issued even if `WITH_HPP_FCL` is not defined. Same for the other macros

Reply:
It is driven by PINOCCHIO_WITH_HPP_FCL which defined and propagated by the Pinocchio project.

Reply:
It seems really weird. So, if `PINOCCHIO_WITH_HPP_FCL` is defined, the `WITH_HPP_FCL` is issued, and at the same time a warning is given that `WITH_HPP_FCL` is deprecated? This does not seem correct

Reply:
Shouln't `PINOCCHIO_WITH_HPP_FCL` and `WITH_HPP_FCL` be switched here?
```
#ifdef WITH_HPP_FCL
  #ifndef PINOCCHIO_WITH_HPP_FCL
    #define PINOCCHIO_WITH_HPP_FCL // for backward compatibility
  #endif
  #ifndef PINOCCHIO_DISABLE_DEPRECATED_MACRO_WARNING
    #pragma message("'WITH_HPP_FCL' is deprecated. Please use PINOCCHIO_WITH_HPP_FCL instead.")
  #endif
```
With the same for all other deprecated macros?

Reply:
Another option that the `PINOCCHIO_DISABLE_DEPRECATED_MACRO_WARNING` could be `PINOCCHIO_ENABLE_COMPATIBILITY_WITH_VERSION_1` and define `WITH_HPP_FCL` if the backward compatibility was requested, without compilation warning.

Reply:
It is Pinocchio that manages PINOCCHIO_WITH_HPP_FCL or WITH_HPP_FCL.
So, if the user has used WITH_HPP_FCL in his project, it has to be defined. This is done here for backward compatibility.

In fact, there is no clean way of making a deprecated Macro as far as I know.

Reply:
Cannot this be handled in CMake, i.e. define WITH_HPP_FCL is backward compatibility is requested ?

Reply:
I see the intended use case now. It is not really clean that the user is forced to issue a macro in order to silence a warning which is not caused by his own code. My feeling is that if a user wants to keep using deprecated macros, it is him that should handle them, not those who switch to the updated version

Reply:
Also, one reason of adding the `PINOCCHIO` prefix was to avoid name clashes. This solution does not solve this problem. Every time `PINOCCHIO_WITH_HPP_FCL` is defined, then `WITH_HPP_FCL` is also defined.

Reply:
The solution from @jmirabel solves this.

Reply:
I also agree with @jmirabel

Reply:
@jmirabel Do you think that `PINOCCHIO_ENABLE_COMPATIBILITY_WITH_VERSION_1` should be defined ON by default?

Reply:
I don't think so. I assumed that people moving from Pinocchio 1 to Pinocchio 2 are ready to dig further into encountered problems. It should be mentioned in the porting notes.

Anyway, for WITH_HPP_FCL, I don't think there will be much problem.